### PR TITLE
Changes to ensure the correct set of symbols are exported

### PR DESCRIPTION
**Pull request by bdonlan**
*Issue #, if available:*

*Description of changes:*

This PR makes a couple of changes to improve the quality of the emitted shared libraries. In particular, we make sure that 'static inline' functions that are part of the public API also have a standalone symbol emitted for FFI purposes. We also turn on `-fvisibility=hidden`; currently, there aren't any symbols affected by this, but there may be some in the future.

The full list of exported symbols (on Darwin) are:

```
0000000000001f40 g       0f SECT   01 0000 [.text] _aws_array_list_back
0000000000002190 g       0f SECT   01 0000 [.text] _aws_array_list_capacity
0000000000001cb0 g       0f SECT   01 0000 [.text] _aws_array_list_clean_up
0000000000001f20 g       0f SECT   01 0000 [.text] _aws_array_list_clear
0000000000007420 g       0f SECT   01 0000 [.text] _aws_array_list_comparator_string
0000000000002060 g       0f SECT   01 0000 [.text] _aws_array_list_copy
0000000000001e40 g       0f SECT   01 0000 [.text] _aws_array_list_front
00000000000021b0 g       0f SECT   01 0000 [.text] _aws_array_list_get_at
00000000000021f0 g       0f SECT   01 0000 [.text] _aws_array_list_get_at_ptr
0000000000001c10 g       0f SECT   01 0000 [.text] _aws_array_list_init_dynamic
0000000000001c80 g       0f SECT   01 0000 [.text] _aws_array_list_init_static
00000000000021a0 g       0f SECT   01 0000 [.text] _aws_array_list_length
0000000000001f80 g       0f SECT   01 0000 [.text] _aws_array_list_pop_back
0000000000001e70 g       0f SECT   01 0000 [.text] _aws_array_list_pop_front
0000000000001ed0 g       0f SECT   01 0000 [.text] _aws_array_list_pop_front_n
0000000000001d00 g       0f SECT   01 0000 [.text] _aws_array_list_push_back
0000000000001d50 g       0f SECT   01 0000 [.text] _aws_array_list_set_at
0000000000001fd0 g       0f SECT   01 0000 [.text] _aws_array_list_shrink_to_fit
00000000000023d0 g       0f SECT   01 0000 [.text] _aws_array_list_sort
0000000000002340 g       0f SECT   01 0000 [.text] _aws_array_list_swap
0000000000002110 g       0f SECT   01 0000 [.text] _aws_array_list_swap_contents
0000000000002fd0 g       0f SECT   01 0000 [.text] _aws_atomic_compare_exchange_int
0000000000002b20 g       0f SECT   01 0000 [.text] _aws_atomic_compare_exchange_int_explicit
0000000000002ff0 g       0f SECT   01 0000 [.text] _aws_atomic_compare_exchange_ptr
0000000000002ba0 g       0f SECT   01 0000 [.text] _aws_atomic_compare_exchange_ptr_explicit
0000000000002fb0 g       0f SECT   01 0000 [.text] _aws_atomic_exchange_int
0000000000002a90 g       0f SECT   01 0000 [.text] _aws_atomic_exchange_int_explicit
0000000000002fc0 g       0f SECT   01 0000 [.text] _aws_atomic_exchange_ptr
0000000000002ad0 g       0f SECT   01 0000 [.text] _aws_atomic_exchange_ptr_explicit
0000000000003010 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_add
0000000000002c20 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_add_explicit
0000000000003040 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_and
0000000000002d90 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_and_explicit
0000000000003060 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_or
0000000000002cc0 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_or_explicit
0000000000003020 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_sub
0000000000002c70 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_sub_explicit
0000000000003080 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_xor
0000000000002e60 g       0f SECT   01 0000 [.text] _aws_atomic_fetch_xor_explicit
0000000000002990 g       0f SECT   01 0000 [.text] _aws_atomic_init_int
00000000000029a0 g       0f SECT   01 0000 [.text] _aws_atomic_init_ptr
0000000000002f70 g       0f SECT   01 0000 [.text] _aws_atomic_load_int
00000000000029b0 g       0f SECT   01 0000 [.text] _aws_atomic_load_int_explicit
0000000000002f80 g       0f SECT   01 0000 [.text] _aws_atomic_load_ptr
00000000000029f0 g       0f SECT   01 0000 [.text] _aws_atomic_load_ptr_explicit
0000000000002f90 g       0f SECT   01 0000 [.text] _aws_atomic_store_int
0000000000002a20 g       0f SECT   01 0000 [.text] _aws_atomic_store_int_explicit
0000000000002fa0 g       0f SECT   01 0000 [.text] _aws_atomic_store_ptr
0000000000002a60 g       0f SECT   01 0000 [.text] _aws_atomic_store_ptr_explicit
0000000000002f30 g       0f SECT   01 0000 [.text] _aws_atomic_thread_fence
0000000000004b10 g       0f SECT   01 0000 [.text] _aws_base64_compute_decoded_len
0000000000004ad0 g       0f SECT   01 0000 [.text] _aws_base64_compute_encoded_len
0000000000004d00 g       0f SECT   01 0000 [.text] _aws_base64_decode
0000000000004b80 g       0f SECT   01 0000 [.text] _aws_base64_encode
0000000000002890 g       0f SECT   01 0000 [.text] _aws_byte_buf_append
0000000000002740 g       0f SECT   01 0000 [.text] _aws_byte_buf_cat
0000000000002440 g       0f SECT   01 0000 [.text] _aws_byte_buf_clean_up
00000000000024c0 g       0f SECT   01 0000 [.text] _aws_byte_buf_clean_up_secure
0000000000002530 g       0f SECT   01 0000 [.text] _aws_byte_buf_eq
0000000000003140 g       0f SECT   01 0000 [.text] _aws_byte_buf_from_array
0000000000003110 g       0f SECT   01 0000 [.text] _aws_byte_buf_from_c_str
00000000000023f0 g       0f SECT   01 0000 [.text] _aws_byte_buf_init
0000000000002570 g       0f SECT   01 0000 [.text] _aws_byte_buf_init_copy
0000000000002490 g       0f SECT   01 0000 [.text] _aws_byte_buf_secure_zero
0000000000002730 g       0f SECT   01 0000 [.text] _aws_byte_buf_split_on_char
00000000000025f0 g       0f SECT   01 0000 [.text] _aws_byte_buf_split_on_char_n
00000000000031a0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_advance
00000000000031d0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_advance_nospec
00000000000028e0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_eq
0000000000002920 g       0f SECT   01 0000 [.text] _aws_byte_cursor_eq_byte_buf
0000000000003170 g       0f SECT   01 0000 [.text] _aws_byte_cursor_from_array
0000000000003160 g       0f SECT   01 0000 [.text] _aws_byte_cursor_from_buf
00000000000041f0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_from_string
0000000000003230 g       0f SECT   01 0000 [.text] _aws_byte_cursor_read
00000000000032a0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_read_and_fill_buffer
0000000000003390 g       0f SECT   01 0000 [.text] _aws_byte_cursor_read_be16
0000000000003400 g       0f SECT   01 0000 [.text] _aws_byte_cursor_read_be32
0000000000003470 g       0f SECT   01 0000 [.text] _aws_byte_cursor_read_be64
0000000000003330 g       0f SECT   01 0000 [.text] _aws_byte_cursor_read_u8
00000000000034e0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write
0000000000003630 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write_be16
00000000000036a0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write_be32
0000000000003710 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write_be64
0000000000003550 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write_from_whole_buffer
0000000000004180 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write_from_whole_string
00000000000035c0 g       0f SECT   01 0000 [.text] _aws_byte_cursor_write_u8
0000000000006240 g       0f SECT   01 0000 [.text] _aws_c_string_eq
0000000000006710 g       0f SECT   01 0000 [.text] _aws_condition_variable_clean_up
00000000000066f0 g       0f SECT   01 0000 [.text] _aws_condition_variable_init
0000000000006760 g       0f SECT   01 0000 [.text] _aws_condition_variable_notify_all
0000000000006720 g       0f SECT   01 0000 [.text] _aws_condition_variable_notify_one
00000000000067a0 g       0f SECT   01 0000 [.text] _aws_condition_variable_wait
00000000000067e0 g       0f SECT   01 0000 [.text] _aws_condition_variable_wait_for
0000000000004790 g       0f SECT   01 0000 [.text] _aws_condition_variable_wait_for_pred
0000000000004740 g       0f SECT   01 0000 [.text] _aws_condition_variable_wait_pred
0000000000004230 g       0f SECT   01 0000 [.text] _aws_default_allocator
0000000000006cc0 g       0f SECT   01 0000 [.text] _aws_default_thread_options
0000000000005060 g       0f SECT   01 0000 [.text] _aws_error_debug_str
0000000000005000 g       0f SECT   01 0000 [.text] _aws_error_lib_name
0000000000004fa0 g       0f SECT   01 0000 [.text] _aws_error_str
0000000000005f20 g       0f SECT   01 0000 [.text] _aws_hash_c_string
0000000000005e10 g       0f SECT   01 0000 [.text] _aws_hash_iter_begin
0000000000005ea0 g       0f SECT   01 0000 [.text] _aws_hash_iter_done
0000000000005eb0 g       0f SECT   01 0000 [.text] _aws_hash_iter_next
00000000000061f0 g       0f SECT   01 0000 [.text] _aws_hash_ptr
00000000000061b0 g       0f SECT   01 0000 [.text] _aws_hash_string
00000000000053b0 g       0f SECT   01 0000 [.text] _aws_hash_table_clean_up
00000000000053f0 g       0f SECT   01 0000 [.text] _aws_hash_table_clear
0000000000005610 g       0f SECT   01 0000 [.text] _aws_hash_table_create
00000000000054f0 g       0f SECT   01 0000 [.text] _aws_hash_table_find
0000000000005d70 g       0f SECT   01 0000 [.text] _aws_hash_table_foreach
00000000000051f0 g       0f SECT   01 0000 [.text] _aws_hash_table_get_entry_count
0000000000005200 g       0f SECT   01 0000 [.text] _aws_hash_table_init
0000000000005ac0 g       0f SECT   01 0000 [.text] _aws_hash_table_put
0000000000005b50 g       0f SECT   01 0000 [.text] _aws_hash_table_remove
0000000000004950 g       0f SECT   01 0000 [.text] _aws_hex_compute_decoded_len
00000000000047e0 g       0f SECT   01 0000 [.text] _aws_hex_compute_encoded_len
0000000000004980 g       0f SECT   01 0000 [.text] _aws_hex_decode
0000000000004800 g       0f SECT   01 0000 [.text] _aws_hex_encode
0000000000006650 g       0f SECT   01 0000 [.text] _aws_high_res_clock_get_ticks
00000000000030f0 g       0f SECT   01 0000 [.text] _aws_hton16
00000000000030d0 g       0f SECT   01 0000 [.text] _aws_hton32
00000000000030b0 g       0f SECT   01 0000 [.text] _aws_hton64
00000000000030a0 g       0f SECT   01 0000 [.text] _aws_is_big_endian
0000000000004f80 g       0f SECT   01 0000 [.text] _aws_last_error
00000000000038c0 g       0f SECT   01 0000 [.text] _aws_linked_list_begin
0000000000003930 g       0f SECT   01 0000 [.text] _aws_linked_list_empty
00000000000038d0 g       0f SECT   01 0000 [.text] _aws_linked_list_end
00000000000039e0 g       0f SECT   01 0000 [.text] _aws_linked_list_front
00000000000038a0 g       0f SECT   01 0000 [.text] _aws_linked_list_init
00000000000038f0 g       0f SECT   01 0000 [.text] _aws_linked_list_insert_after
0000000000003910 g       0f SECT   01 0000 [.text] _aws_linked_list_insert_before
00000000000038e0 g       0f SECT   01 0000 [.text] _aws_linked_list_next
0000000000003880 g       0f SECT   01 0000 [.text] _aws_linked_list_node_reset
0000000000003990 g       0f SECT   01 0000 [.text] _aws_linked_list_pop_back
00000000000039f0 g       0f SECT   01 0000 [.text] _aws_linked_list_pop_front
0000000000003970 g       0f SECT   01 0000 [.text] _aws_linked_list_push_back
00000000000039c0 g       0f SECT   01 0000 [.text] _aws_linked_list_push_front
0000000000003940 g       0f SECT   01 0000 [.text] _aws_linked_list_remove
0000000000003a20 g       0f SECT   01 0000 [.text] _aws_linked_list_swap_contents
00000000000046e0 g       0f SECT   01 0000 [.text] _aws_load_error_strings
0000000000006350 g       0f SECT   01 0000 [.text] _aws_lru_cache_clean_up
00000000000065b0 g       0f SECT   01 0000 [.text] _aws_lru_cache_clear
00000000000063b0 g       0f SECT   01 0000 [.text] _aws_lru_cache_find
0000000000006640 g       0f SECT   01 0000 [.text] _aws_lru_cache_get_element_count
0000000000006620 g       0f SECT   01 0000 [.text] _aws_lru_cache_get_mru_element
0000000000006290 g       0f SECT   01 0000 [.text] _aws_lru_cache_init
0000000000006440 g       0f SECT   01 0000 [.text] _aws_lru_cache_put
0000000000006590 g       0f SECT   01 0000 [.text] _aws_lru_cache_remove
00000000000065c0 g       0f SECT   01 0000 [.text] _aws_lru_cache_use_lru_element
0000000000004240 g       0f SECT   01 0000 [.text] _aws_mem_acquire
0000000000004270 g       0f SECT   01 0000 [.text] _aws_mem_acquire_many
0000000000004430 g       0f SECT   01 0000 [.text] _aws_mem_realloc
0000000000004420 g       0f SECT   01 0000 [.text] _aws_mem_release
0000000000003ba0 g       0f SECT   01 0000 [.text] _aws_mul_size_checked
0000000000003b60 g       0f SECT   01 0000 [.text] _aws_mul_size_saturating
0000000000003ac0 g       0f SECT   01 0000 [.text] _aws_mul_u32_checked
0000000000003aa0 g       0f SECT   01 0000 [.text] _aws_mul_u32_saturating
0000000000003b20 g       0f SECT   01 0000 [.text] _aws_mul_u64_checked
0000000000003ae0 g       0f SECT   01 0000 [.text] _aws_mul_u64_saturating
0000000000006890 g       0f SECT   01 0000 [.text] _aws_mutex_clean_up
00000000000068a0 g       0f SECT   01 0000 [.text] _aws_mutex_init
0000000000006a60 g       0f SECT   01 0000 [.text] _aws_mutex_lock
0000000000006b20 g       0f SECT   01 0000 [.text] _aws_mutex_try_lock
0000000000006be0 g       0f SECT   01 0000 [.text] _aws_mutex_unlock
0000000000003180 g       0f SECT   01 0000 [.text] _aws_nospec_mask
0000000000003100 g       0f SECT   01 0000 [.text] _aws_ntoh16
00000000000030e0 g       0f SECT   01 0000 [.text] _aws_ntoh32
00000000000030c0 g       0f SECT   01 0000 [.text] _aws_ntoh64
00000000000072a0 g       0f SECT   01 0000 [.text] _aws_priority_queue_capacity
0000000000006fc0 g       0f SECT   01 0000 [.text] _aws_priority_queue_clean_up
0000000000006f80 g       0f SECT   01 0000 [.text] _aws_priority_queue_init_dynamic
0000000000006fa0 g       0f SECT   01 0000 [.text] _aws_priority_queue_init_static
0000000000007100 g       0f SECT   01 0000 [.text] _aws_priority_queue_pop
0000000000006fd0 g       0f SECT   01 0000 [.text] _aws_priority_queue_push
0000000000007290 g       0f SECT   01 0000 [.text] _aws_priority_queue_size
0000000000007250 g       0f SECT   01 0000 [.text] _aws_priority_queue_top
0000000000006280 g       0f SECT   01 0000 [.text] _aws_ptr_eq
00000000000050c0 g       0f SECT   01 0000 [.text] _aws_raise_error
0000000000003870 g       0f SECT   01 0000 [.text] _aws_read_u16
0000000000003830 g       0f SECT   01 0000 [.text] _aws_read_u24
0000000000003800 g       0f SECT   01 0000 [.text] _aws_read_u32
00000000000037e0 g       0f SECT   01 0000 [.text] _aws_read_u64
00000000000051b0 g       0f SECT   01 0000 [.text] _aws_register_error_info
0000000000005120 g       0f SECT   01 0000 [.text] _aws_reset_error
0000000000005140 g       0f SECT   01 0000 [.text] _aws_restore_error
0000000000003c80 g       0f SECT   01 0000 [.text] _aws_rw_lock_clean_up
0000000000003bc0 g       0f SECT   01 0000 [.text] _aws_rw_lock_init
0000000000003c90 g       0f SECT   01 0000 [.text] _aws_rw_lock_rlock
0000000000003f90 g       0f SECT   01 0000 [.text] _aws_rw_lock_runlock
0000000000003e10 g       0f SECT   01 0000 [.text] _aws_rw_lock_try_rlock
0000000000003ed0 g       0f SECT   01 0000 [.text] _aws_rw_lock_try_wlock
0000000000003d50 g       0f SECT   01 0000 [.text] _aws_rw_lock_wlock
0000000000004050 g       0f SECT   01 0000 [.text] _aws_rw_lock_wunlock
0000000000002970 g       0f SECT   01 0000 [.text] _aws_secure_zero
0000000000005160 g       0f SECT   01 0000 [.text] _aws_set_global_error_handler_fn
0000000000005180 g       0f SECT   01 0000 [.text] _aws_set_thread_local_error_handler_fn
0000000000004110 g       0f SECT   01 0000 [.text] _aws_string_bytes
00000000000073d0 g       0f SECT   01 0000 [.text] _aws_string_compare
0000000000007370 g       0f SECT   01 0000 [.text] _aws_string_destroy
0000000000007390 g       0f SECT   01 0000 [.text] _aws_string_destroy_secure
0000000000006250 g       0f SECT   01 0000 [.text] _aws_string_eq
0000000000004150 g       0f SECT   01 0000 [.text] _aws_string_eq_byte_buf
0000000000004120 g       0f SECT   01 0000 [.text] _aws_string_eq_byte_cursor
0000000000007310 g       0f SECT   01 0000 [.text] _aws_string_new_from_array
00000000000072b0 g       0f SECT   01 0000 [.text] _aws_string_new_from_c_str
00000000000066a0 g       0f SECT   01 0000 [.text] _aws_sys_clock_get_ticks
0000000000006ca0 g       0f SECT   01 0000 [.text] _aws_system_info_processor_count
0000000000004200 g       0f SECT   01 0000 [.text] _aws_task_init
00000000000074e0 g       0f SECT   01 0000 [.text] _aws_task_scheduler_clean_up
0000000000007570 g       0f SECT   01 0000 [.text] _aws_task_scheduler_has_tasks
0000000000007470 g       0f SECT   01 0000 [.text] _aws_task_scheduler_init
00000000000078e0 g       0f SECT   01 0000 [.text] _aws_task_scheduler_run_all
0000000000007860 g       0f SECT   01 0000 [.text] _aws_task_scheduler_schedule_future
0000000000007830 g       0f SECT   01 0000 [.text] _aws_task_scheduler_schedule_now
0000000000006cd0 g       0f SECT   01 0000 [.text] _aws_thread_clean_up
0000000000006f30 g       0f SECT   01 0000 [.text] _aws_thread_current_sleep
0000000000006f20 g       0f SECT   01 0000 [.text] _aws_thread_current_thread_id
0000000000006eb0 g       0f SECT   01 0000 [.text] _aws_thread_get_detach_state
0000000000006ea0 g       0f SECT   01 0000 [.text] _aws_thread_get_id
0000000000006cf0 g       0f SECT   01 0000 [.text] _aws_thread_init
0000000000006ec0 g       0f SECT   01 0000 [.text] _aws_thread_join
0000000000006d10 g       0f SECT   01 0000 [.text] _aws_thread_launch
0000000000003780 g       0f SECT   01 0000 [.text] _aws_timestamp_convert
00000000000046d0 g       0f SECT   01 0000 [.text] _aws_wrapped_cf_allocator_destroy
00000000000044f0 g       0f SECT   01 0000 [.text] _aws_wrapped_cf_allocator_new
0000000000003850 g       0f SECT   01 0000 [.text] _aws_write_u16
0000000000003810 g       0f SECT   01 0000 [.text] _aws_write_u24
00000000000037f0 g       0f SECT   01 0000 [.text] _aws_write_u32
00000000000037d0 g       0f SECT   01 0000 [.text] _aws_write_u64
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/posix/common.inl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <aws/common/common.h>
+
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static inline int aws_private_convert_and_raise_error_code(int error_code) {
```
Comment (ColdenCullen) [inline include/aws/common/posix/common.inl]: Just curious, why isn't this one `AWS_STATIC_IMPL`?

File: include/aws/common/posix/rw_lock.inl [hunk 2]
```diff
@@ -1,5 +1,6 @@
+#include <aws/common/atomics.h>
```
Comment (ColdenCullen) [inline include/aws/common/posix/rw_lock.inl]: This needs to be inside the include guard or clang-format gets all rekt

File: include/aws/common/posix/common.inl (hunk 1 above)
Comment (bdonlan) [reply include/aws/common/posix/common.inl]: This is not a public API, so we don't want to export symbols for it.

Comment (ColdenCullen) [reply include/aws/common/posix/common.inl]: Ah, that makes sense.

**Pull request merged**